lib/ansible/module_utils/facts/network/linux.py

```diff
@@ -147,9 +147,13 @@ def get_interfaces_info(self, ip_path, default_ipv4, default_ipv6):
                 interfaces[device]['perm_macaddress'] = get_file_content(os.path.join(path, 'bonding_slave', 'perm_hwaddr'), default='')
             if os.path.exists(os.path.join(path, 'device')):
                 interfaces[device]['pciid'] = os.path.basename(os.readlink(os.path.join(path, 'device')))
+            if os.path.exists(os.path.join(path, 'duplex')):
+                duplex = get_file_content(os.path.join(path, 'duplex'))
+                if duplex is not None and duplex != 'unknown':
+                    interfaces[device]['duplex'] = duplex
             if os.path.exists(os.path.join(path, 'speed')):
                 speed = get_file_content(os.path.join(path, 'speed'))
-                if speed is not None:
+                if speed is not None and duplex is not None and duplex != 'unknown':
```

Comment:
> In my case, when the duplex display "unknown", the speed will display wrong, it will display like "65535" or a larger number. At this time, i think not show better than show a wrong data even if this error is not caused by our query.

Interesting. I assume ethtool output for those cases does the same?

I'm wondering if that indicates ethtool cant detect the speed? Not sure we can tell when the speed is unknown because link is down, or if it is unknown because the NIC's ethtool support isnt complete. (http://elixir.free-electrons.com/linux/v4.12-rc6/source/net/core/ethtool.c#L558)
Looks like http://elixir.free-electrons.com/linux/v4.12-rc6/source/include/uapi/linux/ethtool.h#L1507 defines unknown speed as decimal -1.

http://elixir.free-electrons.com/linux/v4.12-rc6/source/include/uapi/linux/ethtool.h#L74 implies that the validity of the 'speed' value also depends on autonegotiation setting which further complicates interpretation of the data.

I dont think we should add a ansible specific definition of these values, since that will lose information and introduce ambiguity. It may make sense to also introduce a new field or fact for our higher level interpretation of the results. That would allow tools/users that need the low level info to get while also providing a higher level summary.